odl/trafos/fourier.py

```diff
@@ -1645,6 +1647,211 @@ def inverse(self):
+
+        if len(non_uniform_samples) == 0:
+            raise ValueError('`non_uniform_samples` is empty')
+
+        n_dim = len(self.shape)
+        if not all(len(sample) == n_dim for sample in non_uniform_samples):
+            raise ValueError('One of the samples in `non_uniform_samples` does'
+              ' not have the right dimension')
+
+        if not all(
+                self._is_sample_valid(sample) for sample in non_uniform_samples
+            ):
+            raise ValueError('One of the samples in `non_uniform_samples` is'
+              ' not a float between -0.5 and 0.5')
+
+    def _is_sample_valid(self, sample):
+        """Check whether a single sample is correct.
+
+        Parameters
+        ----------
+        sample : iterable
+            The sample whose location you want to check
+
+        Returns
+        -------
+        bool : Whether the sample contains only floats in [-0.5; 0.5[
+        """
+        floats = all(isinstance(coord, float) for coord in sample)
+        if floats:
+            normalized = all(-0.5 <= coord < 0.5 for coord in sample)
+        else:
+            normalized = False
+        return floats & normalized
+
+    def _normalize(self, x):
+        """Normalize the result of the non uniform FFT.
+
+        Parameters
+        ----------
+        x : `numpy.ndarray`
+            The results of the non uniform FFT
+
+        Returns
+        -------
+        out : `numpy.ndarray`
+            Normalized results of the non uniform FFT
+        """
+        out = x / np.sqrt(self.nfft.M)
+        return out
+
+    @property
+    def adjoint(self):
+        """Adjoint of this instance."""
+        if self.adjoint_class:
+            return self.adjoint_class(
+                shape=self.shape,
+                non_uniform_samples=self.non_uniform_samples,
+            )
+        raise NotImplementedError(
+            "Adjoint not implemented for this non-uniform fourier operator",
+        )
+
+
+class NonUniformFourierTransform(NonUniformFourierTransformBase):
+    """Forward Non uniform Fast Fourier Transform.
+    """
+    def __init__(self, shape, non_uniform_samples):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        shape : tuple
+            The dimensions of the data whose non uniform FFT you want to
+            compute
+        non_uniform_samples : iterable
+            List of the fourier space positions where the coefficients are
+            computed. Each position must be in [-0.5; 0.5[
+        """
+        super(NonUniformFourierTransform, self).__init__(
+            shape=shape,
+            non_uniform_samples=non_uniform_samples,
+            domain=discr_sequence_space(shape, dtype=np.complex128),
+            range=discr_sequence_space(
+                [len(non_uniform_samples)],
+                dtype=np.complex128,
+            ),
+        )
+        self.adjoint_class = NonUniformFourierTransformAdjoint
```

Comment:
Why store this?

Reply:
hmm this way I could mutualise a bit the code for the adjoint between the 2 classes and put it in the base class. Do you think it's overcomplexified?

Reply:
In general it's perfectly fine to factor out common code. But here we're talking about two lines of code maybe? I think it's more readable to inline the code in the two classes. (Also makes storing an attribute obsolete, which is always good.)

Reply:
yes yes it's not that much. I just can't have it in the `__init__` because of the infinite recursion but even so it's not that bad.

Reply:
You can refer to the classes in the `adjoint` property directly. If you worry about efficiency (overhead of creating new operators, shouldn't be too much here, though) you can always cache the adjoint.

Reply:
Yes yes definitely it's [what I did](https://github.com/odlgroup/odl/pull/1488/files?file-filters%5B%5D=.cfg&file-filters%5B%5D=.py#diff-7533defd16313b3c1ec23dd40014de57R144).